# Сервис для чтения книг по подписке

## Цель исследования

Компания купила крупный сервис для чтения книг по подписке. Задача: проанализировать базу данных. В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

## Описание данных
Таблица **books**
содержит данные о книгах:
- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

Таблица **authors**
содержит данные об авторах:
- `author_id` — идентификатор автора;
- `author` — имя автора.

Таблица **publishers**
содержит данные об издательствах:
- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства.

Таблица **ratings**
содержит данные о пользовательских оценках книг:
- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

Таблица **reviews**
содержит данные о пользовательских обзорах на книги:
- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import create_engine
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
db_config['pwd'],
db_config['host'],
db_config['port'],
db_config['db'])
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

Посмотрим на структуру данных, выведем информацию по всем таблицам.

In [2]:
query = ''' SELECT *
              FROM books;                
        '''
pd.io.sql.read_sql(query, con = engine)

book_id  author_id                                              title  \
0          1        546                                       'Salem's Lot   
1          2        465                 1 000 Places to See Before You Die   
2          3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3          4         82  1491: New Revelations of the Americas Before C...   
4          5        125                                               1776   
..       ...        ...                                                ...   
995      996        571           Wyrd Sisters (Discworld  #6; Witches #2)   
996      997        454                        Xenocide (Ender's Saga  #3)   
997      998        201                                    Year of Wonders   
998      999         94                        You Suck (A Love Story  #2)   
999     1000        509  Zen and the Art of Motorcycle Maintenance: An ...   

     num_pages publication_date  publisher_id  
0          594       2005-11-01            93  
1          992       2003-05-22           336  
2          322       2010-12-21           135  
3          541       2006-10-10           309  
4          386       2006-07-04           268  
..         ...              ...           ...  
995        265       2001-02-06           147  
996        592       1996-07-15           297  
997        358       2002-04-30           212  
998        328       2007-01-16           331  
999        540       2006-04-25           143  

[1000 rows x 6 columns]

In [3]:
query = ''' SELECT *
              FROM authors;                
        '''
pd.io.sql.read_sql(query, con = engine)

author_id                                author
0            1                            A.S. Byatt
1            2        Aesop/Laura Harris/Laura Gibbs
2            3                       Agatha Christie
3            4                         Alan Brennert
4            5              Alan Moore/David   Lloyd
..         ...                                   ...
631        632         William Strunk Jr./E.B. White
632        633                           Zadie Smith
633        634                 Zilpha Keatley Snyder
634        635                    Zora Neale Hurston
635        636  Åsne Seierstad/Ingrid Christopherson

[636 rows x 2 columns]

In [4]:
query = ''' SELECT *
              FROM publishers;                
        '''
pd.io.sql.read_sql(query, con = engine)

publisher_id                          publisher
0               1                                Ace
1               2                           Ace Book
2               3                          Ace Books
3               4                      Ace Hardcover
4               5  Addison Wesley Publishing Company
..            ...                                ...
335           336         Workman Publishing Company
336           337                         Wyatt Book
337           338              Yale University Press
338           339                           Yearling
339           340                     Yearling Books

[340 rows x 2 columns]

In [5]:
query = ''' SELECT *
              FROM ratings;                
        '''
pd.io.sql.read_sql(query, con = engine)

rating_id  book_id        username  rating
0             1        1      ryanfranco       4
1             2        1   grantpatricia       2
2             3        1    brandtandrea       5
3             4        2        lorichen       3
4             5        2     mariokeller       2
...         ...      ...             ...     ...
6451       6452     1000  carolrodriguez       4
6452       6453     1000         wendy18       4
6453       6454     1000      jarvispaul       5
6454       6455     1000           zross       2
6455       6456     1000         fharris       5

[6456 rows x 4 columns]

In [6]:
query = ''' SELECT *
              FROM reviews;               
        '''
pd.io.sql.read_sql(query, con = engine)

review_id  book_id        username  \
0             1        1    brandtandrea   
1             2        1      ryanfranco   
2             3        2        lorichen   
3             4        3   johnsonamanda   
4             5        3     scotttamara   
...         ...      ...             ...   
2788       2789      999      martinadam   
2789       2790     1000         wknight   
2790       2791     1000  carolrodriguez   
2791       2792     1000         wendy18   
2792       2793     1000      jarvispaul   

                                                   text  
0     Mention society tell send professor analysis. ...  
1     Foot glass pretty audience hit themselves. Amo...  
2     Listen treat keep worry. Miss husband tax but ...  
3     Finally month interesting blue could nature cu...  
4     Nation purpose heavy give wait song will. List...  
...                                                 ...  
2788  Later hospital turn easy community. Fact same ...  
2789   Change lose answer close pressure. Spend so now.  
2790  Authority go who television entire hair guy po...  
2791  Or western offer wonder ask. More hear phone f...  
2792  Republican staff bit eat material measure plan...  

[2793 rows x 4 columns]

Итак, мы имеем 1000 различных книг 635 авторов, изданных в 339 издательствах.

## Задания
- Посчитать, сколько книг вышло после 1 января 2000 года;
- Для каждой книги посчитать количество обзоров и среднюю оценку;
- Определить издательство, которое выпустило наибольшее число книг толще 50 страниц, таким образом исключая из анализа брошюры;
- Определить автора с самой высокой средней оценкой книг, учитывая только книги с 50 и более оценками;
- Посчитать среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

### Cколько книг вышло после 1 января 2000 года

In [7]:
# Количество книг, вышедших после 1 января 2000 года

query = ''' SELECT 'Кол-во книг, вышедших после 1 января 2000 г' AS about,
                    COUNT(DISTINCT book_id) 
              FROM books
             WHERE publication_date > '2000-01-01';
        '''

pd.io.sql.read_sql(query, con = engine)

about  count
0  Кол-во книг, вышедших после 1 января 2000 г    819

Большая часть книг вышла после этой даты (819 из 1000).

### Для каждой книги посчитать количество обзоров и среднюю оценку

In [8]:
#  количество обзоров и средняя оценка для каждой книги   

query = '''          
            WITH       
                 new_rating AS (            /* посчитали среднюю оценку  */
                                SELECT book_id,
                                       ROUND(AVG(rating),2) AS mean_rating
                                  FROM ratings 
                                 GROUP BY book_id
                               ),
                         
                 new_review AS (            /* посчитали кол-во ревью  */
                                SELECT book_id,
                                       COUNT(review_id) AS review_number
                                  FROM reviews
                                 GROUP BY book_id
                               )
                              
          SELECT b.title,
                 rev.review_number,
                 rat.mean_rating
                 
            FROM books AS b
            LEFT OUTER JOIN new_rating AS rat ON b.book_id = rat.book_id
            LEFT OUTER JOIN new_review AS rev ON b.book_id = rev.book_id
           WHERE b.num_pages > 50
                ;            
        '''

pd.io.sql.read_sql(query, con = engine)

title  review_number  \
0            The Body in the Library (Miss Marple  #3)            2.0   
1                                            Galápagos            2.0   
2                             A Tree Grows in Brooklyn            5.0   
3    Undaunted Courage: The Pioneering First Missio...            2.0   
4                                          The Prophet            4.0   
..                                                 ...            ...   
987   The Cat in the Hat and Other Dr. Seuss Favorites            NaN   
988    Anne Rice's The Vampire Lestat: A Graphic Novel            NaN   
989                          Essential Tales and Poems            NaN   
990                               Leonardo's Notebooks            NaN   
991                            The Natural Way to Draw            NaN   

     mean_rating  
0           4.50  
1           4.50  
2           4.25  
3           4.00  
4           4.29  
..           ...  
987         5.00  
988         3.67  
989         4.00  
990         4.00  
991         3.00  

[992 rows x 3 columns]

###  Определить издательство, которое выпустило наибольшее число книг толще 50 страниц, таким образом исключая из анализа брошюры

In [9]:
#  Издательство, которое выпустило наибольшее число книг (толще 50 страниц)

query = ''' SELECT p.publisher,
                   COUNT(p.publisher_id) AS books_number  
                   
              FROM books AS b
              LEFT OUTER JOIN publishers AS p ON b.publisher_id = p.publisher_id
             WHERE b.num_pages > 50
             GROUP BY p.publisher_id
             ORDER BY COUNT(p.publisher_id) DESC
             LIMIT 1;
        '''

pd.io.sql.read_sql(query, con = engine)

publisher  books_number
0  Penguin Books            42

Больше всего книг из этой базы данных выпустило издательство Penguin Books - 42 книги.

### Определить автора с самой высокой средней оценкой книг, учитывая только книги с 50 и более оценками

In [10]:
#  Автор с самой высокой средней оценкой книги (учитены только книги с 50 и более оценками)

query = '''   WITH    
                   select_book AS (            /* Сразу нашли одну книгу из ">=50 оценок" и max AVG */
                                   SELECT book_id,
                                          AVG(rating) AS mean_rating  /* Нашли средний рейтинг книг */
                                     FROM ratings 
                                    GROUP BY book_id
                                   HAVING COUNT(rating_id) >= 50 /* Оставили только книги с 50 и более оценками */ 
                                    ORDER BY AVG(rating) DESC     /* Сортировка по убыванию */
                                    LIMIT 1                       /* Один (первый) элемент - MAX значение */
                                  )
            SELECT a.author,
                   sb.mean_rating
           
              FROM select_book AS sb
              LEFT OUTER JOIN books AS b ON sb.book_id  = b.book_id
              LEFT OUTER JOIN authors AS a ON b.author_id  = a.author_id;
              
        '''

pd.io.sql.read_sql(query, con = engine)

author  mean_rating
0  J.K. Rowling/Mary GrandPré     4.414634

Это J.K. Rowling/Mary GrandPré

### Посчитать среднее количество обзоров от пользователей, которые поставили больше 50 оценок

In [11]:
# Среднее количество обзоров от пользователей, которые поставили больше 50 оценок

query = ''' WITH       
                 new_rating AS (            /* отобрали пользователей с кол-вом оценок > 50) */
                                SELECT username
                                  FROM ratings 
                                 GROUP BY username
                                HAVING COUNT(book_id) > 50
                               ),
                               
                 new_review AS (            /* посчитали кол-во ревью у пользователей */
                               SELECT username,
                                      COUNT(book_id) AS reviews_number
                                 FROM reviews
                                GROUP BY username
                               )
                              
          SELECT ROUND(AVG(nrev.reviews_number),2) AS mean_review   
                                            /*  Нашли среднее арифм. в сводной таблице и округлили */
            FROM new_review AS nrev
           INNER JOIN new_rating AS nrat ON nrat.username = nrev.username; 
        '''
pd.io.sql.read_sql(query, con = engine)

mean_review
0        24.33

Наиболее активные пользователи (критики) оставляют в среднем порядка 24-25 отзывов на прочитанные книги.